tamr_toolbox/low_latency/llm.py

```diff
@@ -0,0 +1,208 @@
+            list containing matcched data
+    Raises:
+        ValueError: if match type is not "records" or "clusters", or if batch_size is non-positive
+        RuntimeError: if query fails
+    """
+
+    result_dict = defaultdict(lambda: [])  # dict which defaults to empty list to hold results
+
+    url = f"/api/v1/projects/{project_name}:match?type={type}"
+
+    # Set up keys to read results
+    if type == "records":
+        record_key = "queryRecordId"
+        prob_key = "matchProbability"
+    elif type == "clusters":
+        record_key = "entityId"
+        prob_key = "avgMatchProb"
+    else:
+        raise ValueError(f"Unsupported match type {type}.")
+
+    # Convert single record to list for processing
+    if isinstance(records, Dict):
+        records = [records]
+
+    # Check batch size and set if not supplied
+    if batch_size is None:
+        batch_size = len(records)
+        if batch_size == 0:
+            LOGGER.warn("No input supplied to llm_query -- returning empty result.")
+            return result_dict
+    elif batch_size <= 0:
+        raise ValueError(f"Batch size must be non-negative: received {batch_size}")
+
+    # Split into batches and convert to LLM query format
+    for j in range(0, len(records), batch_size):
+        json_recs = _prepare_json(records[j : j + batch_size], pKey=pKey, offset=j)
+        response = match_client.post(url, json=json_recs)
+
+        # Process responses
+        if response.ok:
+            if response.content == b"":  # handle null response
+                continue
+
+            # If data was found, decode, identify source record, and add match to corresponding
+            # index in the list of lists of results
+            for resp_block in response.content.decode("utf-8").split("\n"):
+                if resp_block:
+                    result = json.loads(resp_block)
+                    index = int(result[record_key]) if pKey is None else result[record_key]
+
+                    if max_num_matches and len(result_dict[index]) >= max_num_matches:
+                        continue
+
+                    if min_match_prob and result[prob_key] < min_match_prob:
+                        continue
+
+                    result_dict[index].append(result)
+
+        else:
+            message = f"LLM query failed: {response.content}"
+            LOGGER.error(message)
+            raise RuntimeError(message)
+
+    return result_dict
+
+
+def _prepare_json(
+    records: List[JsonDict], *, pKey: Union[str, None], offset: int
+) -> List[JsonDict]:
+    """
+    Put records into JSON format expected by LLM endpoint
+
+    Args:
+        records: list of records to match
+        pKey: a primary key for the data; if not None, this must be a field in the input records
+        offset: offset to apply to generated integer `recordId` -- this is necessary for batching
+    Returns:
+        List of formatted records
+    Raises:
```

Comment:
Docstring formatting:
```suggestion
        offset: offset to apply to generated integer `recordId` -- this is necessary for batching
    
    Returns:
        List of formatted records
    
    Raises:
```